### DISCUSSION TITLE
How to fetch an image from the server/api?

### DISCUSSION THREAD

**Discussion by durbanitas** [nuxt/nuxt#16061] [Questions]
I'm struggeling to use a simple fetch to get an image from an endpoint.

#### Current approach

```js
// test.vue
const img = await $fetch('/api/image')
console.log(img); // 
```
```js
// server/api/image.js
export default defineEventHandler(async (event) => {
  const url = 'https://images-assets.nasa.gov/image/PIA19512/PIA19512~thumb.jpg'

  return await fetch(url, {
    headers: {
      mode: 'cors',
      'access-control-allow-origin': '*',
      'content-type': 'image/jpeg'
    }
  })
    .then(res => {
    console.log(res);  // url: 'https://images-assets.nasa.gov/image/PIA19512/PIA19512~thumb.jpg',
                       // status: 200,
                       // statusText: 'OK',
                       // [...]
    return res
  })
})
```
I see a 200 response but the nitro console tells me:

> [nitro] [dev] [uncaughtException] TypeError [ERR_INVALID_ARG_TYPE]: The "chunk" argument must be of type string or an instance of Buffer. Received an instance of Response

No if I am trying to consume the body I got the following error:

```js
// server/api/image.js
export default defineEventHandler(async (event) => {
  const url = 'https://images-assets.nasa.gov/image/PIA19512/PIA19512~thumb.jpg'

  return await fetch(url, {
    headers: {
      mode: 'cors',
      'access-control-allow-origin': '*',
      'content-type': 'image/jpeg'
    }
  })
    .then(res => {
      return res.blob()
    })
    .then(resData => {
      console.log(resData); // Blob {}
    })
})
```
```
[nuxt] [request error] body used already for: https://images-assets.nasa.gov/image/PIA19512/PIA19512~thumb.jpg
  at consumeBody (/E:/Code/ratios/node_modules/node-fetch-native/dist/chunks/abort-controller.mjs:4955:9)
  at Response.arrayBuffer (/E:/Code/ratios/node_modules/node-fetch-native/dist/chunks/abort-controller.mjs:4865:50)
  at Response.blob (/E:/Code/ratios/node_modules/node-fetch-native/dist/chunks/abort-controller.mjs:4894:26)
  at /E:/Code/ratios/.nuxt/dev/index.mjs:472:16
  at processTicksAndRejections (internal/process/task_queues.js:95:5)
  at async /E:/Code/ratios/.nuxt/dev/index.mjs:464:10
  at async /E:/Code/ratios/node_modules/h3/dist/index.mjs:420:19
  at async Server.nodeHandler (/E:/Code/ratios/node_modules/h3/dist/index.mjs:370:7)
```

---

#### Second try and logging the response inside the endpoint

![403 error](https://imgur.com/a3C2JrE.png)

Is it possible that I am doing something fundamentally wrong with my approach? 

I can't figure it out at the moment how to transform the chunk argument into a string or buffer. Do you have any ideas?

**Comment by tylerforesthauser** [comment]
Not sure this is a complete solution but hopefully it points you in the right direction. I believe the key is using arrayBuffer for your fetch response type and converting that to Buffer for the final response.

```js
import { send } from 'h3'

export default defineEventHandler(async (event) => {
  const config = useRuntimeConfig()

  const { params: { id } } = event.context

  const response = await $fetch.raw(`${config.apiUrl}/assets/${id}`, { responseType: 'arrayBuffer' })

  const contentType = response.headers.get('Content-Type')

  return send(event, Buffer.from(_data), contentType)
})
```

**Reply by durbanitas** [reply]
Thank you for showing me a different approach. I'm currently getting a `403`-error when I'm logging `response`. Adding the cors headers aren't successful.

**Comment by dangngocbinh** [comment]
It should work

const data = fs.readFileSync("/app/example2.png");
event.res.setHeader("Content-Type", "image/png");
event.res.write(Buffer.from(data));
event.res.end();

**Comment by MrSunshyne** [comment]
I have a similar question but slightly different approach. How could i use nuxt-image module to fetch the image and transform it in server/api ?

i.e
`server/api/article.ts`
```ts 
export default defineEventHandler(async (event) => {
  const $img = useImage();

  const post = await getPostFromSdk('my-id');

  let newImage = await $img(post.coverImage, {
        width: 1920,
        height: 1080,
        format: "webp",
  });

  return {
    title: post.title,
    optimizedCover: newImage
  }
})
```

I'm not even sure if it makes sense to use useImage here, or any other composables from nuxt..

**Comment by Croodo** [comment]
The image URL you are trying to fetch is protected by CloudFront that's why you are getting a 403 status code.
To debug your code you can try to fetch an image from a different domain that is not protected by CloudFront or CloudFlare.
